Combobox Pattern and Examples: Convert to support draft ARIA 1.2 specification

mcking65 opened the pull request:
Resolves #1250, resolves #1244.

### Preview link

[Preview work in progress in the compare branch for this PR](https://raw.githack.com/w3c/aria-practices/issue1244-aria1.2-combobox-conversion/aria-practices.html#combobox)

### Review checklist

* [x] [Editorial review](https://github.com/w3c/aria-practices/wiki/Pull-Request-Review-Process#editorial_review) by @carmacleod, @charmarkk, @jongund
* [x] [Functional review](https://github.com/w3c/aria-practices/wiki/Pull-Request-Review-Process#functional_review) by @carmacleod, @smhigley 
* [x] [Visual design review](https://github.com/w3c/aria-practices/wiki/Pull-Request-Review-Process#design_review) by @zcorpan - https://github.com/w3c/aria-practices/issues/1263
* [x] [Windows Accessibility review](https://github.com/w3c/aria-practices/wiki/Pull-Request-Review-Process#accessibility_review) by @carmacleod, @smhigley 
* [x] [Mac Accessibility review](https://github.com/w3c/aria-practices/wiki/Pull-Request-Review-Process#accessibility_review) by @smhigley 
* [ ] [Code review](https://github.com/w3c/aria-practices/wiki/Pull-Request-Review-Process#code_review) by @smhigley 
* [x] [Test review](https://github.com/w3c/aria-practices/wiki/Pull-Request-Review-Process#test_review) by @spectranaut


<!--
    This comment and the below content is programatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/w3c/aria-practices/pull/1255.html" title="Last updated on Nov 14, 2019, 7:39 PM UTC (f535354)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/aria-practices/1255/2b1c5b1...f535354.html" title="Last updated on Nov 14, 2019, 7:39 PM UTC (f535354)">Diff</a>

jongund commented:
@mcking65 Added the gird combobox example to the branch

jongund commented:
@mcking65 I updated the regression tests for the modified combobox examples for both, inline and none

mcking65 commented:
@jongund, thank you, thank you for the updates!!! I finished the editorial updates to the grid combobox example page. So, now, except for the tests, all 4 examples are ready for review.

Do you have time to finish work on the tests?

Tomorrow, Simon will work on the pattern revisions, and then the entire PR will be ready for review.

zcorpan commented on examples/combobox/js/listbox.js:
```diff
@@ -7,7 +7,7 @@ var Listbox = function (domNode, comboboxObj) {
 
   // Check whether domNode is a DOM element
-  if (!(domNode instanceof Element)) {
+  if (!domNode instanceof Element) {
```
```suggestion
  if (!(domNode instanceof Element)) {
```
This needs the inner parantheses, since `!` binds more strongly than `instanceof`. That is, the above is equivalent to

```
if ((!domNode) instanceof Element) {
```

zcorpan replied:
I noticed now that this also fails in the linter:
```
/home/travis/build/w3c/aria-practices/examples/combobox/js/listbox.js
  10:7  error  Unexpected negating the left operand of 'instanceof' operator  no-unsafe-negation
```

zcorpan replied:
Fixed in https://github.com/w3c/aria-practices/pull/1255/commits/29e94eee65a532be3dcbbf0e4c9bad5083584454

zcorpan commented on examples/combobox/js/listboxOption.js:
```diff
@@ -2,7 +2,7 @@
 *   https://www.w3.org/Consortium/Legal/2015/copyright-software-and-document
 */
-var ListboxOption = function (domNode, listboxObj) {
+var Option = function (domNode, listboxObj) {
```
`Option` is a standard constructor.

https://html.spec.whatwg.org/multipage/form-elements.html#the-option-element

This would override the standard constructor, which works fine so long as there isn't other code that uses it.

Until we rewrite examples to use modules, at least, we should avoid global variables with the same names as standard constructors.

zcorpan replied:
I noticed now that this also fails in the linter:
```
/home/travis/build/w3c/aria-practices/examples/combobox/js/listboxOption.js
  5:5  error  'Option' is already defined as a built-in global variable  no-redeclare
```

zcorpan replied:
Fixed in https://github.com/w3c/aria-practices/pull/1255/commits/29e94eee65a532be3dcbbf0e4c9bad5083584454

zcorpan commented on aria-practices.html:
```diff
@@ -712,31 +712,30 @@ <h4>WAI-ARIA Roles, States, and Properties</h4>
     <section class="widget" id="combobox">
       <h3>Combo Box</h3>
```
The spelling of combo box (2 words) vs combobox (1 word) is inconsistent. I prefer combobox.

zcorpan replied:
Changed in https://github.com/w3c/aria-practices/pull/1255/commits/f82e215c70ec084dde67eb3c60eef4e34bd09415

zcorpan commented on aria-practices.html:
```diff
@@ -712,31 +712,30 @@ <h4>WAI-ARIA Roles, States, and Properties</h4>
       <p>
-        A <a href="#combobox" class="role-reference">combobox</a> is a widget made up of the combination of two distinct elements:
-        1) a single-line textbox, and 2) an associated pop-up element for helping users set the value of the textbox.
+        A <a href="#combobox" class="role-reference">combobox</a> is a widget that controls a pop-up element for helping users set the value of the combobox.
```
The spelling of pop-up vs popup is inconsistent. I prefer popup (without a dash).

zcorpan replied:
Changed in https://github.com/w3c/aria-practices/pull/1255/commits/621b26c6c89c79c5219ae061b2736d0f87df8855

zcorpan commented on aria-practices.html:
```diff
@@ -984,75 +975,55 @@ <h5>Tree Popup Keyboard Interaction</h5>
+            That is, <code>aria-haspopup</code> is set to <code>grid</code>, <code>tree</code>, or <code>dialog</code>.
+            Note that elements with role <code>combobox</code> have an implicit <code>aria-haspopup</code> value of <code>listbox</code>.
           </li>
+          <li>When the popup is visible, the combobox element has <a href="#aria-controls" class="property-reference">aria-controls</a> set to a value that refers to the popup element.</li>
```
Should we remove "When the popup is visible"? If I understand correctly, the popup element is always present in the DOM, and `aria-controls` should be present always. The popup is just styled with `display: none` when hidden.

zcorpan replied:
Changed in https://github.com/w3c/aria-practices/pull/1255/commits/e6e8499495b3084fa3d131c88ebca414e0164eda

mcking65 replied:
The actual requirement is aria-controls only needs to be present when the popup is visible. There is discussion about what browsers should expose when aria-controls refers to something not present or not visible. That said, it is not an error to refer to an element that is not visible.

mcking65 replied:
I revised this text to make the actual requirement more clear but specify that it is valid to refer to an element that is not visible.

zcorpan commented on aria-practices.html:
```diff
@@ -984,75 +975,55 @@ <h5>Tree Popup Keyboard Interaction</h5>
+          <li>When a descendant of a listbox, grid, or tree popup is focused, DOM focus remains on the combobox and the combobox has <a href="#aria-activedescendant" class="property-reference">aria-activedescendant</a> set to a value that refers to the focused element within the popup.</li>
+          <li>For a combobox that controls a listbox, grid, or tree popup, when a suggested value is visually indicated as the currently selected value, the <code>option</code>, <code>gridcell</code>, <code>row</code>, or <code>treeitem</code> containing that value has <a href="#aria-selected" class="state-reference">aria-selected</a> set to <code>true</code>.</li>
           <li>
             If the combobox has a visible label, the element with role combobox has <a href="#aria-labelledby" class="property-reference">aria-labelledby</a> set to a value that refers to the labelling element.
```
The combobox can be an HTML element (`input`) that can be labeled with the HTML `label` element. Should we say that `label` element is preferred?

zcorpan replied:
Changed in https://github.com/w3c/aria-practices/pull/1255/commits/2b926c8cd16970ac8f438f6b4c3380ec7ba08917

zcorpan commented on aria-practices.html:
```diff
@@ -771,39 +772,31 @@ <h3>Combo Box</h3>
-      </p>
 
       <section class="notoc">
         <h4>Examples</h4>
```
There is currently no example of a combobox that is not editable. We should add such an example.

carmacleod replied:
Agree. Please add thoughts to this issue: https://github.com/w3c/aria-practices/issues/1026

zcorpan commented:
@mcking65 and @jongund I would appreciate your review on my changes. Thanks!

Also see https://github.com/w3c/aria-practices/issues/1250#issuecomment-552560140

zcorpan commented:
We should update the tests as well.

jongund commented:
@mcking65 Why did you change the test ids grid-combo.html?  Everything was working before the changes.  I work on fixing the test cases to use the new test ids.

mcking65 commented:
@jongund asked:

> @mcking65 Why did you change the test ids grid-combo.html? Everything was working before the changes. I work on fixing the test cases to use the new test ids.

Jon, I'm sorry ... I didn't mean to screw that up. I thought I made only changes where the test ID didn't match what needed to be tested. Technically, since we are using a text input, the ones that include textbox in the ID are fine, and I could have left that part of the ID alone.

jongund commented:
@mcking65 I updated the grid combobox regression test files

jongund commented:
@mcking65 

In the ARIA practices section for combobox "WAI-ARIA Roles, States, and Properties"

Item 1
> If the combobox element is editable, it has a value for aria-multiline of false. Note that the default value of aria-multiline is false.

1. The design pattern mentions the use of `aria-multiline=false` if the combobox is editable, but none of our editable comnbobox examples defines the `aria-multiline` property.  What is even more confusing is that the practices say the default value is `false`, so does it need to be defined or not?  If not why do we say anything.   
2. We also do not say what to do if the values in the combobox can only be from the options in the popup.

Item 2
> When a combobox receives focus, DOM focus is placed on the combobox element

Does this statement make any sense now that the textbox has the combobox role?

carmacleod commented:
Agree with @jongund regarding confusion around mentioning `aria-multiline`. It makes the reader wonder if it's possible to have a combobox with `aria-multiline="true"`? Does combo allow selecting and typing paragraphs? Maybe "Select your favorite cat quote"?

carmacleod commented on aria-practices.html:
```diff
@@ -710,100 +710,102 @@ <h4>WAI-ARIA Roles, States, and Properties</h4>
-        A <a href="#combobox" class="role-reference">combobox</a> is a widget made up of the combination of two distinct elements:
-        1) a single-line textbox, and 2) an associated pop-up element for helping users set the value of the textbox.
+        A <a href="#combobox" class="role-reference">combobox</a> is an input widget with an associated popup that enables users to select a value for the combobox from a collection of possible values.
+        In some implementations, the popup presents allowed values, while in other implementations, the popup presents suggested values , and users may either select one of the suggestions or type a value.
```
delete space before comma

mcking65 replied:
got it, thanks.

carmacleod commented on examples/grid/LayoutGrids.html:
```diff
@@ -240,7 +240,7 @@ <h4 id="ex3_search_results_label">Search Results for "W3C WAI-ARIA"</h4>
               tabindex="-1" role="gridcell"
             >Jul 21, 2016 - This version of ARIA 1.1 changes the aria-kbdshortcuts property to
-              aria-keyshortcuts , expands roles that can be used in combo boxes, adds a ...</span>
+              aria-keyshortcuts , expands roles that can be used in comboboxes, adds a ...</span>
```
delete space before comma

zcorpan commented:
@jongund I think both of those can be removed.

The ARIA spec has the requirement about `aria-multiline`, but including it in aria-practices is more confusing than not mentioning it at all.

charmarkk commented on aria-practices.html:
```diff
@@ -710,100 +710,102 @@ <h4>WAI-ARIA Roles, States, and Properties</h4>
+        Alternatively, if a combobox supports text input, it is referred to as editable.
+        An editable combobox may either allow users to input any arbitrary value, or it may restrict its value to a discrete set of allowed values, in which case typing input serves to filter suggestions presented in the popup.
+      </p>
+      <p>The popup is hidden by default, i.e., the default state is collapsed. The conditions that trigger expansion -- display of the popup --are specific to each implementation. Some possible conditions that trigger expansion include:</p>
```
```suggestion
      <p>The popup is hidden by default, i.e., the default state is collapsed. The conditions that trigger expansion -- display of the popup -- are specific to each implementation. Some possible conditions that trigger expansion include:</p>
```
Added space after em dash following "display of the popup" to separate it from "are specifically."

carmacleod replied:
add space: "--are"

charmarkk commented on aria-practices.html:
```diff
@@ -710,100 +710,102 @@ <h4>WAI-ARIA Roles, States, and Properties</h4>
       <ol>
         <li>
-          <strong>No autocomplete:</strong> When the popup is triggered, the suggested values it contains are the same regardless of the characters typed in the textbox.
+          <strong>No autocomplete:</strong> The combobox is editable, and  when the popup is triggered, the suggested values it contains are the same regardless of the characters typed in the combobox.
```
```suggestion
          <strong>No autocomplete:</strong> The combobox is editable, and when the popup is triggered, the suggested values it contains are the same regardless of the characters typed in the combobox.
```
Remove space after first "and" following "editable."

charmarkk commented on aria-practices.html:
```diff
@@ -710,100 +710,102 @@ <h4>WAI-ARIA Roles, States, and Properties</h4>
         <li>
-          <strong>No autocomplete:</strong> When the popup is triggered, the suggested values it contains are the same regardless of the characters typed in the textbox.
+          <strong>No autocomplete:</strong> The combobox is editable, and  when the popup is triggered, the suggested values it contains are the same regardless of the characters typed in the combobox.
           For example, the popup suggests a set of recently entered values, and the suggestions  do not change as the user types.
```
```suggestion
          For example, the popup suggests a set of recently entered values, and the suggestions do not change as the user types.
```
Remove space after "suggestions."

charmarkk commented on aria-practices.html:
```diff
@@ -710,100 +710,102 @@ <h4>WAI-ARIA Roles, States, and Properties</h4>
         <li>
-          <strong>List autocomplete with manual selection:</strong> When the popup is triggered, it presents suggested values that complete or logically correspond to the characters typed in the textbox.
-          The character string the user has typed will become the value of the textbox unless the user selects a value in the popup.
+          <strong>List autocomplete with manual selection:</strong> When the popup is triggered, it presents suggested values. If the combobox is editable,  the suggested values complete or logically correspond to the characters typed in the combobox.
```
```suggestion
          <strong>List autocomplete with manual selection:</strong> When the popup is triggered, it presents suggested values. If the combobox is editable, the suggested values complete or logically correspond to the characters typed in the combobox.
```
Remove space after comma following "editable."

charmarkk commented on aria-practices.html:
```diff
@@ -710,100 +710,102 @@ <h4>WAI-ARIA Roles, States, and Properties</h4>
+        Two other widgets that are also visually compact and enable users to make a single choice from a set of discrete choices are <a href="#Listbox">listbox</a> and <a href="#menubutton">menu button</a>.
+        One feature that distinguishes combobox from both listbox and menu button is that the user's choice can be presented as a value in an editable field, which gives users the ability to select some or all of the value for copying to the clipboard.
+        Comboboxes and menu buttons can be implemented so users can explore the set of allowed choices without losing a previously made choice.
+        That is, users can navigate the set of available choices in a combobox popup or menu and then press <kbd>escape</kbd>, which closes the popup or menu without changing previous input.
```
```suggestion
        That is, users can navigate the set of available choices in a combobox popup or menu and then press <kbd>Escape</kbd>, which closes the popup or menu without changing previous input.
```
Capitalize "Escape" for consistency.

carmacleod replied:
also, add "the" after "previous" so that it reads, "changing the previous input".

charmarkk commented on aria-practices.html:
```diff
@@ -812,81 +814,90 @@ <h5>Textbox Keyboard Interaction</h5>
+        <div class="note">
+          <p>Standard single line text editing keys appropriate for the device platform:</p>
+          <ol>
+            <li>include keys for input, cursor movement, selection, and text manipulation.</li>
```
```suggestion
            <li>Include keys for input, cursor movement, selection, and text manipulation.</li>
```
Capitalize "include" for consistency with other `<li>`s.

charmarkk commented on aria-practices.html:
```diff
@@ -812,81 +814,90 @@ <h5>Textbox Keyboard Interaction</h5>
+          <ol>
+            <li>include keys for input, cursor movement, selection, and text manipulation.</li>
+            <li>Standard key assignments for editing functions depend on the device operating system.</li>
+            <li>The most robust approach for providing text editing functions is to rely on browsers, which  supply them for HTML inputs with type text and for elements with the <code>contenteditable</code> HTML attribute.</li>
```
```suggestion
            <li>The most robust approach for providing text editing functions is to rely on browsers, which supply them for HTML inputs with type text and for elements with the <code>contenteditable</code> HTML attribute.</li>
```
Remove space between "which supply."

charmarkk commented on aria-practices.html:
```diff
@@ -812,81 +814,90 @@ <h5>Textbox Keyboard Interaction</h5>
         <ul>
-          <li><kbd>Enter</kbd>: Accepts the focused option in the listbox by closing the popup and placing the accepted value in the textbox with the input cursor at the end of the value.</li>
-          <li><kbd>Escape</kbd>: Closes the popup and returns focus to the textbox. Optionally, clears the contents of the textbox.</li>
+          <li><kbd>Enter</kbd>: Accepts the focused option in the listbox  by closing the popup, placing the accepted value in the combobox, and if the combobox is editable,  placing the input cursor at the end of the value.</li>
```
```suggestion
          <li><kbd>Enter</kbd>: Accepts the focused option in the listbox by closing the popup, placing the accepted value in the combobox, and if the combobox is editable,  placing the input cursor at the end of the value.</li>
```
Remove space between "listbox by."

charmarkk commented on aria-practices.html:
```diff
@@ -812,81 +814,90 @@ <h5>Textbox Keyboard Interaction</h5>
-          <li><kbd>Up Arrow</kbd>: Moves focus to and selects the previous option. If focus is on the first option, either returns focus to the textbox or does nothing.</li>
-          <li><kbd>Home</kbd> (Optional): Either moves focus to and selects the first option or returns focus to the textbox and places the cursor on the first character.</li>
-          <li><kbd>End</kbd> (Optional): Either moves focus to the last option or returns focus to the textbox and places the cursor after the last character.</li>
+          <li><kbd>Home</kbd> (Optional): Either moves focus to and selects the first option or, if the combobox is editable,  returns focus to the combobox and places the cursor on the first character.</li>
```
```suggestion
          <li><kbd>Home</kbd> (Optional): Either moves focus to and selects the first option or, if the combobox is editable, returns focus to the combobox and places the cursor on the first character.</li>
```
Remove space after "editable."

carmacleod commented:
See example bug reported by @LaurenceRLewis in https://github.com/w3c/aria-practices/issues/1261#issuecomment-553581609 ... every instance of `aria-haspopup="true"` in the example code and in the Role, Property, State section of the examples should either be changed to `aria-haspopup="listbox"` or deleted completely.

charmarkk commented on aria-practices.html:
```diff
@@ -984,91 +1003,82 @@ <h5>Tree Popup Keyboard Interaction</h5>
-              <li><code>none</code>: When the popup is displayed, the suggested values it contains are the same regardless of the characters typed in the textbox.</li>
-              <li><code>list</code>: When the popup is triggered, it presents suggested values that complete or logically correspond to the characters typed in the textbox.</li>
+              <li><code>none</code>: When the popup is displayed, the suggested values it contains are the same regardless of the characters typed in the combobox.</li>
+              <li><code>list</code>: When the popup is triggered, it presents suggested values. If the combobox is editable,  the values complete or logically correspond to the characters typed in the combobox.</li>
```
```suggestion
              <li><code>list</code>: When the popup is triggered, it presents suggested values. If the combobox is editable, the values complete or logically correspond to the characters typed in the combobox.</li>
```
Remove space after "editable."

charmarkk commented on aria-practices.html:
```diff
@@ -6583,12 +6593,12 @@ <h2>Structural Roles</h2>
     </p>
     <p>
-      some host languages, such as HTML, include elements that express the same  semantics as an ARIA role. 
+      some host languages, such as HTML, include elements that express the same  semantics as an ARIA role.
```
```suggestion
      some host languages, such as HTML, include elements that express the same semantics as an ARIA role.
```
Remove space after "same."

carmacleod commented on aria-practices.html:
```diff
@@ -710,100 +710,102 @@ <h4>WAI-ARIA Roles, States, and Properties</h4>
+        The one that most shapes interaction is text input.
+        Some comboboxes allow users to type and edit text in the combobox and others do not.
+        If a combobox does not support text input, it is referred to as select-only, meaning the only way users can set a value is by selecting a value in the popup.
+        For example, in some browsers, an HTML <code>select</code> element with <code>size="1"</code> is presented to assistive technologies as a combobox.
```
consider changing "as a combobox" to "as a select-only combobox" to tie this sentence in to the flow of the discussion.

charmarkk replied:
Is this not just saying, effectively, that a `select` with `size="1"` is announced as a combobox by AT? In other words, AT doesn't announce it as "select-only," right? That's how I read it at least.

mcking65 replied:
Mark, correct, screen readers present selects as a plain combobox.

carmacleod commented on aria-practices.html:
```diff
@@ -710,100 +710,102 @@ <h4>WAI-ARIA Roles, States, and Properties</h4>
+        <li>It is displayed when the <kbd>Down Arrow</kbd> key is pressed or the <q>Open</q> button is activated. Optionally, if the combobox is editable and contains a string that does not match an allowed value, expansion does not occur.</li>
+        <li>It is displayed when the combobox receives focus even if the combobox is editable and empty.</li>
+        <li>If the combobox is editable, the popup is displayed only if a certain number of characters are typed in the combobox and those characters match some portion of one of the suggested values.</li>
+        <li>If the combobox is editable, it is displayed if the value of the combobox is altered in a way that creates one or more partial matches to a suggested value.</li>
```
This last point (and how it differs from the one preceding it) is hard to parse.
i.e. what does "altered in a way" mean? Typing, pasting, or selecting from the list?
How is this different from the previous point? Is the key difference "only if a certain number of characters are typed"?

mcking65 replied:
Agree, deleting the last one.

spectranaut commented:
Tests look good to me! I just notice one test that was marked as failing that has previously been fixed.

zcorpan commented:
I've done visual design review in https://github.com/w3c/aria-practices/issues/1263.

> Our examples are not on par with modern design systems/component frameworks.

carmacleod commented on aria-practices.html:
```diff
@@ -812,81 +814,90 @@ <h5>Textbox Keyboard Interaction</h5>
+          </li>
+          <li>Printable Characters:
+            <ul>
+              <li>If the combobox is editable, type characters in the combobox. Note that some implementations may regard certain characters as invalid and prevent their input.</li>
```
```suggestion
              <li>If the combobox is editable, types characters in the combobox. Note that some implementations may regard certain characters as invalid and prevent their input.</li>
```

carmacleod commented on aria-practices.html:
```diff
@@ -812,81 +814,90 @@ <h5>Textbox Keyboard Interaction</h5>
         <h5>Grid Popup Keyboard Interaction</h5>
         <p>
           In a grid popup, each suggested value may be represented by either a single cell or an entire row.
           See notes below for how this aspect of grid design effects the keyboard interaction design and the way that selection moves in response to focus movements.
```
```suggestion
          See notes below for how this aspect of grid design affects the keyboard interaction design and the way that selection moves in response to focus movements.
```

carmacleod commented on aria-practices.html:
```diff
@@ -899,21 +910,24 @@ <h5>Grid Popup Keyboard Interaction</h5>
-          <li>The grid allows only one suggested value to be selected at a time for the textbox value.</li>
+          <li>DOM Focus is maintained on the combobox and the assistive technology focus is moved within the grid using <code>aria-activedescendant</code> as described in <a href="#kbd_focus_activedescendant">Managing Focus in Composites Using aria-activedescendant.</a></li>
+          <li>The grid allows only one suggested value to be selected at a time for the combobox value.</li>
           <li>In a grid popup, each suggested value may be represented by either a single cell or an entire row. This aspect of design effects focus and selection movement:
```
```suggestion
          <li>In a grid popup, each suggested value may be represented by either a single cell or an entire row. This aspect of design affects focus and selection movement:
```

carmacleod commented on aria-practices.html:
```diff
@@ -899,21 +910,24 @@ <h5>Grid Popup Keyboard Interaction</h5>
+              <li>If the combobox is editable, returns focus to the combobox and places the cursor after the last character.</li>
             </ul>
           </li>
           <li><kbd>Control + Home</kbd> (optional): moves focus to the first row.</li>
```
```suggestion
          <li><kbd>Control + Home</kbd> (Optional): moves focus to the first row.</li>
```

carmacleod commented on aria-practices.html:
```diff
@@ -899,21 +910,24 @@ <h5>Grid Popup Keyboard Interaction</h5>
           </li>
           <li><kbd>Control + Home</kbd> (optional): moves focus to the first row.</li>
           <li><kbd>Control + End</kbd> (Optional): moves focus to the last row.</li>
+          <li>Any printable character: If the combobox is editable, returns the focus to the combobox without closing the popup and types the character.</li>
```
```suggestion
          <li>Any printable character: If the combobox is editable, returns the focus to the combobox without closing the popup and types the character into the combobox.</li>
```

smhigley commented on examples/combobox/grid-combo.html:
```diff
@@ -77,13 +76,13 @@ <h2 id="ex_label">Example</h2>
       </div>
```
Suggest adding `aria-readonly="true"` to the grid, to prevent each individual grid cell announcing it is editable (as happens in NVDA + FF)

mcking65 replied:
This is an NVDA bug. It would not align with the ARIA spec to add aria-readonly. There is actually no such thing as a editable grid. You can have an editable gridcell in a spreadshett type application where the cell content is directly editable, but the grid itself is never editable. This is explained in the spec.

smhigley commented on examples/combobox/js/combobox-list.js:
```diff
@@ -57,7 +57,7 @@ ComboboxList.prototype.init = function () {
 
```
There's an error in this file in how the listbox handles the blur event --

line 349 calls `this.removeVisualFocusAll();`, which sets `listbox.hasFocus` to true, which prevents the listbox from closing. The end result is that the listbox does not close when you click outside of it.

carmacleod commented on aria-practices.html:
```diff
@@ -961,10 +974,16 @@ <h5>Tree Popup Keyboard Interaction</h5>
           <li><kbd>Up Arrow</kbd>: Moves focus to the previous node that is focusable without opening or closing a node and selects it if it is selectable.</li>
           <li><kbd>Home</kbd>: Moves focus to the first node in the tree without opening or closing a node and selects it if it is selectable.</li>
```
```suggestion
          <li><kbd>Home</kbd>: Moves focus to the first focusable node in the tree without opening or closing a node and selects it if it is selectable.</li>
```

carmacleod commented on aria-practices.html:
```diff
@@ -961,10 +974,16 @@ <h5>Tree Popup Keyboard Interaction</h5>
           <li><kbd>Up Arrow</kbd>: Moves focus to the previous node that is focusable without opening or closing a node and selects it if it is selectable.</li>
           <li><kbd>Home</kbd>: Moves focus to the first node in the tree without opening or closing a node and selects it if it is selectable.</li>
           <li><kbd>End</kbd>: Moves focus to the last node in the tree that is focusable without opening a node and selects it if it is selectable.</li>
```
```suggestion
          <li><kbd>End</kbd>: Moves focus to the last focusable node in the tree without opening or closing a node and selects it if it is selectable.</li>
```

The prose for Home/End should be exactly the same except for first/last.

carmacleod commented on aria-practices.html:
```diff
@@ -984,91 +1003,82 @@ <h5>Tree Popup Keyboard Interaction</h5>
+              In ARIA 1.0, the combobox referenced its popup with <a href="#aria-owns" class="property-reference">aria-owns</a> instead of <a href="#aria-controls" class="property-reference">aria-controls</a>.
+              While user agents might support comboboxes with <a href="#aria-owns" class="property-reference">aria-owns</a> for backwards compatibility with legacy content, it is strongly recommended that authors use <a href="#aria-controls" class="property-reference">aria-controls</a>.
+            </li>
+            <li>When referring to the roles, states, and properties documentation for the below list of patterns used for popups, keep in mind that a combobox is a single-select widget where selection always follows focus in the popup.</li>
```
```suggestion
            <li>When referring to the roles, states, and properties documentation for the below list of patterns used for popups, keep in mind that a combobox is a single-select widget where selection follows focus in the popup.</li>
```

Maybe delete "always", because selection does not always follow focus in tree popup.

carmacleod commented on examples/combobox/combobox-autocomplete-both.html:
```diff
@@ -196,7 +191,7 @@ <h3 id="kbd_label_listbox">Listbox Popup</h3>
         <strong>NOTE:</strong> When visual focus is in the listbox, DOM focus remains on the textbox and the value of <code>aria-activedescendant</code> on the textbox is set to a value that refers to the listbox option that is visually indicated as focused.
```
For Up arrow, change "last" to "previous" as follows:
> If the listbox is displayed and a suggestion is selected, moves visual focus to the last suggested value.

to:
> If the listbox is displayed and a suggestion is selected, moves visual focus to the previous suggested value.

mcking65 replied:
That's not what happens when focus is in the textbox ... this table is for textbox. If it is in the textbox, up takes you to last suggestion

carmacleod replied:
Nope; with focus in textbox, up arrow goes to previous suggestion in list. :)
Steps:
- focus textbox
- type down arrow 6 times; you will be on California
- type up arrow once; you will be on Arkansas, not Wyoming

JAWS-test commented:
> IE: none of the listbox comboboxes work at all, with or without screen readers running

* IE without screenreader: This was also true for the old examples and seems to be a CSS problem: The list entries are not displayed. I suspect this is easy to fix. 
* IE with JAWS: JAWS + IE 11 does not support any `combobox` pattern (ARIA 1.0, ARIA 1.1., ARIA 1.2). There is no solution for this (https://github.com/FreedomScientific/VFO-standards-support/issues/221). I am in favor of including a clear warning that `role=combobox` must be avoided in IE 11 and other elements should be used (`<select>`, input field with autocomplete etc).

carmacleod commented:
Functional issues:
- I agree with @smhigley that Escape should not clear the text input if the popup is open. The user just wants to close the popup at that point - they do not want their previous value to be erased without their consent. Typing Escape with the popup closed can clear the text input. Fixing this includes rewriting the keyboard interaction prose for Escape in the pattern and example.
The [Combobox Keyboard Interaction](https://raw.githack.com/w3c/aria-practices/issue1244-aria1.2-combobox-conversion/aria-practices.html#combobox-keyboard-interaction) in the pattern correctly says:
    > Escape: Dismisses the popup if it is visible. Optionally, if the popup is hidden before Escape is pressed, clears the combobox.

    But then the [Listbox Popup Keyboard Interaction](https://raw.githack.com/w3c/aria-practices/issue1244-aria1.2-combobox-conversion/aria-practices.html#listbox-popup-keyboard-interaction), [Grid](https://raw.githack.com/w3c/aria-practices/issue1244-aria1.2-combobox-conversion/aria-practices.html#grid-popup-keyboard-interaction), and [Tree](https://raw.githack.com/w3c/aria-practices/issue1244-aria1.2-combobox-conversion/aria-practices.html#tree-popup-keyboard-interaction) change it for some reason, to say:
    > Escape: Closes the popup and returns focus to the combobox. Optionally, if the combobox is editable, clears the contents of the combobox.

    Then, the [Dialog](https://raw.githack.com/w3c/aria-practices/issue1244-aria1.2-combobox-conversion/aria-practices.html#dialog-popup-keyboard-interaction) changes it yet again to be optional:
    > Cancel out of the dialog, e.g., press Escape or activate the cancel button in the dialog. Canceling either returns focus to the text box without changing the combobox value or returns focus to the combobox and clears the combobox.

- Also agree that down arrow should open the popup in the grid example.

Additional functional issues:
- the black hover outline appears when hovering over what looks like empty space off to the right of the button (but is actually the combo's label element). This feels unexpected.

- I somehow managed to cause the popup to show only "Utah" while the combobox value was "Wyoming" (the last value in the list - which I had not selected). This occurred on 2 separate occasions when returning to the open browser tab. I don't think this should be possible. :)
![proof of Wyoming-Utah bug](https://user-images.githubusercontent.com/3331913/68826080-67714d00-066a-11ea-9e0e-ccc0e211a825.png)

mcking65 commented:
JAWS-test wrote:

> > IE: none of the listbox comboboxes work at all, with or without screen readers running
> 
> * IE without screenreader: This was also true for the old examples. The list entries are empty (also in the source code). Probably a JS problem.
> * IE with JAWS: JAWS + IE 11 does not support any `combobox` pattern (ARIA 1.0, ARIA 1.1., ARIA 1.2). There is no solution for this ([FreedomScientific/VFO-standards-support#221](https://github.com/FreedomScientific/VFO-standards-support/issues/221)). I am in favor of including a clear warning that `role=combobox` must be avoided in IE 11 and other elements should be used (`<select>`, input field with autocomplete etc).

I believe this is solveable if Freedom Scientific wishes to solve it. I think it is related to aria-activedescendant. @accdc probably knows the root cause.

It is likely APG will soon drop IE support. However, given the popularity of comboboxes, there may be a good case for making these examples operational in IE if FS does want to make them work in IE.

mcking65 commented:
Sarah, thank you for the fabulous review.

@smhigley wrote:

> * NVDA + Firefox: reads the page title four times almost every time I change options using the arrow key with the listbox open (this may be related to the fact that all option elements are mutated every time one changes selection; I'd be happy to investigate the root cause further). This is a code bug, not an issue with the pattern, as it did not repro in a different implementation.
> * JAWS + Firefox: would sporadically re-announce the listbox when arrowing through options. Seems likely related to the above

Opened #1264 and scheduled to resolve in next milestone.

> * IE: none of the listbox comboboxes work at all, with or without screen readers running

I didn't open an issue for this since we previously discussed dropping IE support in the new year.

> * IE: the grid combobox does not work with JAWS; options are not read when arrowed to.

Known bug in JAWS support for grids.

> * macOS + VoiceOver: the suggested "open" keyboard command does not work (Caps + Space), though Option + down arrow does
> * macOS + VoiceOver: although everything but Caps+ Space worked for me in 10.14, @jnurthen reported that 10.15 did not work; someone else with Catalina should probably officially test this.

My system is still 10.4.6, and it works for me too. We will need to continues this testing prior to APG 1.2 R1.

> * iOS + VoiceOver: the "open" button did not work (didn't open the listbox, and sent the cursor back to the text input). I'd suggest either making the open button `aria-hidden="true" tabindex="-1"`, or removing it altogether.

Surprising suggestion. How would touch users expand/collapse the combobox if it is not accessible?

This is a widget that we should make touch compatible before APG 1.2 release. Opened #1267.

> * Windows high contrast mode: the selected option within the listbox is not visible

Added this to issue #1263.

> General functionality notes and issues:
> 
> * The listbox/grid popup does not close when clicking outside of it (detailed comment on why in the `combobox-list.js` file)

Opened #1265 and scheduled for next milestone.

> * Escape clears the text input (this has been mentioned before). It also means that if you have a currently selected item and open the listbox, it's impossible to dismiss the listbox and restore the original selection with escape

Added to #1066.

> * Grid combobox: it seems odd that this combobox (and only this one) does not let you open the popup and see all options without first typing something
> * This isn't fully a bug, but I find it odd that the list scrolls every time you navigate to a new option, even if that option was already in view before the programmatic scroll

Opened #1266 and schedule for next milestone.

mcking65 commented:
Carolyn,thank you for the awesome reviews!!

Except for the 2 where I commented otherwise, I merged all your suggestions.

@carmacleod wrote:

> Functional issues:
> 
> * I agree with @smhigley that Escape should not clear the text input if the popup is open. The user just wants to close the popup at that point - they do not want their previous value to be erased without their consent. Typing Escape with the popup closed can clear the text input. Fixing this includes rewriting the keyboard interaction prose for Escape in the pattern and example.

Added this to #1066.

> * the black hover outline appears when hovering over what looks like empty space off to the right of the button (but is actually the combo's label element). This feels unexpected.

Is this something that could be addressed in #1263? Or added to one of the other open issues in the combobox project?

> * I somehow managed to cause the popup to show only "Utah" while the combobox value was "Wyoming" (the last value in the list - which I had not selected). This occurred on 2 separate occasions when returning to the open browser tab. I don't think this should be possible. :)
>   ![proof of Wyoming-Utah bug](https://user-images.githubusercontent.com/3331913/68826080-67714d00-066a-11ea-9e0e-ccc0e211a825.png)

Opened #1268 and scheduled in next milestone.

JAWS-test commented:
> I believe this is solveable if Freedom Scientific wishes to solve it. 

I'm afraid that won't happen because the problem has been there for many years and not a single improvement has happened since. 

> I think it is related to aria-activedescendant.

The problem has nothing to do with `aria-activedescendant` (this works fine now, it was just a problem in the past). The problem is that `role=combobox` is not recognized by JAWS. So it's not the list entries that are affected, but the element itself.

mcking65 commented:
@jongund wrote:

> In the ARIA practices section for combobox "WAI-ARIA Roles, States, and Properties"
> 
> Item 1
> 
> > If the combobox element is editable, it has a value for aria-multiline of false. Note that the default value of aria-multiline is false.
> 
> 1. The design pattern mentions the use of `aria-multiline=false` if the combobox is editable, but none of our editable comnbobox examples defines the `aria-multiline` property.  What is even more confusing is that the practices say the default value is `false`, so does it need to be defined or not?  If not why do we say anything.

Removed it in commit dd6f6c0.

> 2. We also do not say what to do if the values in the combobox can only be from the options in the popup.

The very first bullet says that the element that displays the value has role combobox. That's all that is needed. However, I do think that we might eventually benefit from a property something like "aria-valuefrom" that would take an ID and explicitly define the source of the value for the combobox.

> Item 2
> 
> > When a combobox receives focus, DOM focus is placed on the combobox element
> 
> Does this statement make any sense now that the textbox has the combobox role?

Certainly ... we need focus on that element. Obviously, it gets it by default if you use html input type text, but you would need to explicitly make the element focusable if it were not an element that is focusable by default.

mcking65 commented:
@@charmarkk, thank you very much for the thorough review!! I believe I have merged all your suggestions.

mcking65 commented:
@carmacleod wrote:

> See example bug reported by @LaurenceRLewis in [#1261 (comment)](https://github.com/w3c/aria-practices/issues/1261#issuecomment-553581609) ... every instance of `aria-haspopup="true"` in the example code and in the Role, Property, State section of the examples should either be changed to `aria-haspopup="listbox"` or deleted completely.

Removed with commit e51ad04.

@spectranaut, This will need to be changed in the regression tests.

carmacleod commented:
> > * the black hover outline appears when hovering over what looks like empty space off to the right of the button (but is actually the combo's label element). This feels unexpected.
> 
> Is this something that could be addressed in #1263? Or added to one of the other open issues in the combobox project?

Added to #1263 by @zcorpan.

carmacleod commented:
#### VoiceOver+Chrome on macOS Catalina 10.15.1

- focusing the combobox and then typing down arrow: all 3 listbox examples say the list item twice, although the first time is cut short. This is most apparent in the [autocomplete=none example](https://raw.githack.com/w3c/aria-practices/issue1244-aria1.2-combobox-conversion/examples/combobox/combobox-autocomplete-none.html) for "w-weather, ch-cheap flights to NY".

- the [autocomplete=both example](https://raw.githack.com/w3c/aria-practices/issue1244-aria1.2-combobox-conversion/examples/combobox/combobox-autocomplete-both.html) has some very weird behavior where sometimes only a part of the previously-selected list item is spoken as you arrow through the list. For example: focus the combo and type down arrow, and Alaska is selected and spoken (even though the first item is Alabama); type down arrow again, and VO says "lask", down arrow again to hear "merican Samoa Arizona". If you wait long enough before typing down arrow again, it may synchronize so that the selection matches what is spoken. But then down arrow again results in "District of Columbi". It's pretty unusable.

- `ctrl+alt+space` (i.e. VO+space) does open the popup in the listbox (not grid) examples in Chrome (but does not focus the first item)

#### VoiceOver+Safari on macOS Catalina 10.15.1

- VO does not read the list items in the [autocomplete=list](https://raw.githack.com/w3c/aria-practices/issue1244-aria1.2-combobox-conversion/examples/combobox/combobox-autocomplete-list.html) or [autocomplete=none](https://raw.githack.com/w3c/aria-practices/issue1244-aria1.2-combobox-conversion/examples/combobox/combobox-autocomplete-none.html) or [grid](https://raw.githack.com/w3c/aria-practices/issue1244-aria1.2-combobox-conversion/examples/combobox/grid-combo.html) examples in Safari. These 3 examples are completely unusable in VO+Safari on Catalina.

- However, VO _does_ read list items in the [autocomplete=both](https://raw.githack.com/w3c/aria-practices/issue1244-aria1.2-combobox-conversion/examples/combobox/combobox-autocomplete-both.html) example

- VO+space does not open the list in Safari (for any example)

mcking65 commented:
@carmacleod wrote:

> #### VoiceOver+Chrome on macOS Catalina 10.15.1
> * focusing the combobox and then typing down arrow: all 3 listbox examples say the list item twice, although the first time is cut short. This is most apparent in the [autocomplete=none example](https://raw.githack.com/w3c/aria-practices/issue1244-aria1.2-combobox-conversion/examples/combobox/combobox-autocomplete-none.html) for "w-weather, ch-cheap flights to NY".
> * the [autocomplete=both example](https://raw.githack.com/w3c/aria-practices/issue1244-aria1.2-combobox-conversion/examples/combobox/combobox-autocomplete-both.html) has some very weird behavior where sometimes only a part of the previously-selected list item is spoken as you arrow through the list. For example: focus the combo and type down arrow, and Alaska is selected and spoken (even though the first item is Alabama); type down arrow again, and VO says "lask", down arrow again to hear "merican Samoa Arizona". If you wait long enough before typing down arrow again, it may synchronize so that the selection matches what is spoken. But then down arrow again results in "District of Columbi". It's pretty unusable.

Both these behaviors smell like bugs in our code and are also probably related to #1264. @spectranaut is working on #1264 today.

> * `ctrl+alt+space` (i.e. VO+space) does open the popup in the listbox (not grid) examples in Chrome (but does not focus the first item)

Interesting, confusing.

> #### VoiceOver+Safari on macOS Catalina 10.15.1
> * VO does not read the list items in the [autocomplete=list](https://raw.githack.com/w3c/aria-practices/issue1244-aria1.2-combobox-conversion/examples/combobox/combobox-autocomplete-list.html) or [autocomplete=none](https://raw.githack.com/w3c/aria-practices/issue1244-aria1.2-combobox-conversion/examples/combobox/combobox-autocomplete-none.html) or [grid](https://raw.githack.com/w3c/aria-practices/issue1244-aria1.2-combobox-conversion/examples/combobox/grid-combo.html) examples in Safari. These 3 examples are completely unusable in VO+Safari on Catalina.

The fact that they do work in 10.14.6 does not necessarily implicate Safari, but kind of points in that direction. @accdc has some different implementations of the same pattern we could test with 10.15. Apple may have implemented overly restrictive processing of aria-activedescendant, overlooking the fact that the ARIA spec supports activedescendant on a text input or combobox that controls an element that also supports activedescendant.

> * However, VO _does_ read list items in the [autocomplete=both](https://raw.githack.com/w3c/aria-practices/issue1244-aria1.2-combobox-conversion/examples/combobox/combobox-autocomplete-both.html) example

This is almost certainly because of the inline selection.

> * VO+space does not open the list in Safari (for any example)

VO-space is supposed to be equivalent to a mouse click on the element in the VO cursor. Given that, does VO announcing that VO-space should open the popup make sense?

jongund commented:
@mcking65 do we distinguish the term "focus" from "DOM focus" in the practices?

> Item 2
> 
> When a combobox receives focus, DOM focus is placed on the combobox element
> 
> Does this statement make any sense now that the textbox has the combobox role?
> 
> Certainly ... we need focus on that element. Obviously, it gets it by default if you use html input type text, but you would need to explicitly make the element focusable if it were not an element that is focusable by default.
>

charmarkk commented:
> > * VO+space does not open the list in Safari (for any example)
> 
> VO-space is supposed to be equivalent to a mouse click on the element in the VO cursor. Given that, does VO announcing that VO-space should open the popup make sense?

I think that does make sense, Matt.
